oasis-reference/src/main/kotlin/com/empowerops/volition/ref_oasis/OptimizerEndPoint.kt

```diff
@@ -0,0 +1,271 @@
+package com.empowerops.volition.ref_oasis
+
+import com.empowerops.volition.dto.*
+import io.grpc.stub.StreamObserver
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.selects.select
+import org.funktionale.either.Either
+import sun.reflect.generics.reflectiveObjects.NotImplementedException
+import java.time.Duration
+
+sealed class SimResult {
+    data class Success(val name: String, val result: Map<String, Double>) : SimResult()
+    data class TimeOut(val name: String) : SimResult()
+    data class TimeOutFailure(val name: String, val exception: String) : SimResult()
+    data class Failure(val name: String, val exception: String) : SimResult()
+}
+
+class OptimizerEndpoint(
+        private val modelService: ModelService
+) : OptimizerGrpc.OptimizerImplBase(), OptimizationControl {
+    var state = State.Idle
+
+    enum class State {
+        Idle,
+        Running
+    }
+
+    override fun changeNodeName(request: NodeNameChangeCommandDTO, responseObserver: StreamObserver<NodeNameChangeResponseDTO>) = responseObserver.consume {
+        val newName = request.newName
+        val newNameSimulation = modelService.simByName[newName]
+        val oldName = request.oldName
+        val target = modelService.simByName[oldName]
+        if (newNameSimulation == null && target != null) {
+            modelService.renameSim(target, newName, oldName)
+            NodeNameChangeResponseDTO.newBuilder().setChanged(true).build()
+        } else {
+            NodeNameChangeResponseDTO.newBuilder().setChanged(false).build()
+        }
+    }
+
+    override fun register(request: RegistrationCommandDTO, responseObserver: StreamObserver<OASISQueryDTO>) {
+        if (request.name in modelService.simByName.keys) {
+            return
+        }
+        modelService.addNewSim(request.name!!, Simulation(emptyList(), emptyList(), "", responseObserver, Channel(1), Channel(1), Channel(1)))
```

Comment:
Please use `Channel(RENDEZVOUS)` instead of `Channel(1)`.

Reply:
RENDEZVOUS is not 1, I haven't fully explore what the difference are but do you still want me to keep it as ArrayChannel(1) which is Channel(1) will return.

Reply:
Ah you're right, `1` is an array-channel of size 1. 

In this way the behavior is that the first call to any API endpoint returns instantly (because its buffer is at size 0). 

If another API call is made to the same endpoint before the previous one has even started being processed, it is suspended until the first one starts being processed, at which point the previous element is removed from the channel, the second one takes its place, and the API endpoint returns.

With `RENDEZVOUS` the result would be that you would get suspended on the first call until that message starts being processed. I think this is a better behaviour, and its hard to understand how it would introduce a bug.

So yes, please change this to RENDEZVOUS. It should functionally stay the same, but various API calls will likely take a few milliseconds longer.

Reply:
Will do